src/simplewallet/simplewallet.cpp

```diff
@@ -130,10 +130,25 @@ namespace
 
   const command_line::arg_descriptor< std::vector<std::string> > arg_command = {"command", ""};
 
+#ifdef HAVE_READLINE
+  class simplewallet_suspend_readline: public rdln::suspend_readline
+  {
+  public:
+    simplewallet_suspend_readline(): rdln::suspend_readline() { boost::lock_guard<boost::mutex> lock(mutex); in_password = true; }
+    virtual ~simplewallet_suspend_readline() { boost::lock_guard<boost::mutex> lock(mutex); in_password = false; }
+    static bool is_in_password() { boost::lock_guard<boost::mutex> lock(mutex); return in_password; }
+  private:
+    static boost::mutex mutex;
+    static bool in_password;
```

Comment:
I used a std::atomic at first, but I had trouble. I don't think I really care about tiny cache improvements in something that happens in relation to user input and signals though :) Nonetheless, I'm curious to know more about "the cachelin doesn't have to be dirtied" if you don't mind. You mean less code using an atomic, or something else ?

Reply:
Acquiring a lock requires modifying some memory location associated with that lock. Internally the system must wait for all cores across all cpus for exclusive access to that cacheline before proceeding. Reading an atomic variable requires no changes to the cacheline before proceeding. The only exception is if you used a speculative lock instruction which is highly unlikely to ever be used for `boost::mutex` on x86-64 platforms due to its limitations (it would break lots of existing code).

Reply:
Let me walk that back a bit ... I'm not sure that the atomic was highlighting a race in the existing code, because I haven't seen a clear description as to the problem. The PR description states that sometimes the terminal is left with echoing off. If that is what you have observed, why is that happening? The read_tty functions are `noexcept` with only C calls, so it should never exit without turning the echo back on, unless an immediate termination is being done (in that function or elsewhere).

Reply:
The problem was not one of race, but one of the signal triggering a wallet exit while the input was set to noecho. The password read thread was waiting till enter to finish and let the program end, which was also confusing.

Reply:
The PR description says the console was left in hidden mode. Is that the behavior, or blocking program termination until enter is pressed? And how would this fix the issue referenced?

Reply:
Both.And it fixes it by ignoring the signal when in a password entry.

Reply:
Why is the fix only applied to readline builds? What is happening in the signal handler during readline builds to cause this issue? Think about it this way, there is a call to a C function `getchar()` between the enabling and disabling of the command line echo. How is the enable echo function call being skipped? Is an exception thrown in that C function? Or is the process immediately terminating? Because the issue is almost certainly immediate program termination, which could be an issue in non-readline builds (since we have not determined the source of the program termination). If this "fixes" it, then something in the `SIGINT` handler can cause program termination unexpectedly.

A slightly better fix would be to use `sigprocmask` before disabling the echo. This way if any signal is sent to the process - even signals with default program termination - the process will not process them until the echo is turned back on. This will help alleviate the echo being disabled issue, but doesn't fix some woes in the signal handler or the block until enter issue.

The issue causing immediate program termination might be a [self join](https://github.com/monero-project/monero/blob/master/contrib/epee/include/console_handler.h#L117). If the OS chooses the thread associated with that object to handle the signal, it will ultimately try to join itself. The call _probably_ throws then immediately terminates when the stack unwind hits the OS. This would cause the program to terminate while inside of the `getchar()` function, preventing the echo enabling.

[There is also a destruction race that starts here](https://github.com/monero-project/monero/blob/master/src/simplewallet/simplewallet.cpp#L5717). That call tells a thread to exit the command processing loop, which eventually [returns to here](https://github.com/monero-project/monero/blob/master/src/simplewallet/simplewallet.cpp#L6825). If the OS uses a thread other than the main one or the one created in the console handler, the signal handler could still be executing when the main thread is cleaning up the stack for `main` itself. I'm not sure what happens here, but it cannot be right because the process has many exit routines to perform even after everything in the `main` function has been destroyed (i.e. the process is still active).

Fixing the blocking issue is a little more involved. Switching to `read` without signal blocking is not a fix - the handler can execute between the check for shutdown and the call to `read`. Use of `select` has the same problem. Using `pselect` with the sigmask field would be closer, but the handler would have to be fixed. Another intriguing option is to block signals, then use `signalfd` (linux specific) and the equivalent in BSD to have `select` notify of when a signal arrives.

Reply:
> Why is the fix only applied to readline builds? 

Agreed. It's not a readline issue, rather an issue in password_container. I have been trying to fix (unsuccessfully) and could do with some help.

Reply:
@vtnerd after reading through your detailed comment (thanks btw) a couple of times, I think you are totally correct - and this issue is far more involved than first impressions. It also explains why I was having issues modifying read_from_tty using various select semantics. I never had a crash, but either echo didn't start / restore correctly.

Reply:
That's a very good point :)
But anyway, it'll be fixed by the read_from_tty fix anyway. I'll close it, it's just generating lots of stuff for a patch that's already been found to be worse than another.

Reply:
@vtnerd OK I got the signal blocking working to fix the echo restore issue. Now am hitting up against:
> The issue causing immediate program termination might be a self join. If the OS chooses the thread associated with that object to handle the signal, it will ultimately try to join itself. The call probably throws then immediately terminates when the stack unwind hits the OS. 

```
Wallet password: libc++abi.dylib: terminating with uncaught exception of type boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::thread_resource_error> >: boost thread: trying joining itself: Resource deadlock avoided
Abort trap: 6


frame #8: 0x00000001017d54d7 monero-wallet-cli`void boost::throw_exception<boost::thread_resource_error>(e=0x000070000007fff0) + 119 at throw_exception.hpp:69
frame #9: 0x000000010179aded monero-wallet-cli`boost::thread::join(this=0x00007fff5e4fec40) + 125 at thread.hpp:773
frame #10: 0x00000001017fe6bf monero-wallet-cli`epee::async_stdin_reader::stop(this=0x00007fff5e4fec40) + 287 at console_handler.h:117
```